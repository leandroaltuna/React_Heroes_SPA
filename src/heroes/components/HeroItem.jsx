import { Link } from 'react-router-dom';


const CharactersByHero = ({ alter_ego, characters }) => {
    if ( alter_ego === characters ) return(<></>);
    return <p>{ characters }</p>;
}


export const HeroItem = ({ 
    id,
    superhero,
    publisher,
    alter_ego,
    first_appearance,
    characters,   
 }) => {

    const heroImgUrl = `/heroes/${ id }.jpg`;
    // const heroImgUrl = `/assets/heroes/${ id }.jpg`;
    // const charactersByHero = ( <p>{ characters }</p> );

    return (
        <>
            <div className="col mb-1 animate__animated animate__fadeIn">
                <div className="card">

                    <div className="row no-gutters">
                        
                        <div className="col-4">
                            <img 
                                className="card-img"
                                src={ heroImgUrl } 
                                alt={ superhero } 
                            />
                        </div>

                        <div className="col-8">
                            <div className="card-body">

                                <h5 className="card-title">{ superhero }</h5>
                                <p className="card-text">{ alter_ego }</p>
                                {/* {
                                    ( alter_ego !== characters ) && charactersByHero
                                } */}
                                <CharactersByHero 
                                    alter_ego={ alter_ego } 
                                    characters={ characters } 
                                />
                                
                                <p className="card-text">
                                    <small className="text-muted">{ first_appearance }</small>
                                </p>

                                <Link to={ `/hero/${ id }` }>
                                    more...
                                </Link>

                            </div>
                        </div>

                    </div>

                </div>
            </div>
        </>
    )

}
