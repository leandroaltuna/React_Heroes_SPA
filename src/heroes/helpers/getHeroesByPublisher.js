import { heroes } from '../data/heroes';


export const getHeroesByPublisher = ( publisher ) => {

    const validPublishers = ['DC Comics', 'Marvel Comics'];

    if ( !validPublishers.includes( publisher ) ) {
        throw new Error( `${ publisher } is not a valid publisher` );
    }

    const heroList = heroes.filter( ( hero ) => hero.publisher === publisher );

    return heroList;

}