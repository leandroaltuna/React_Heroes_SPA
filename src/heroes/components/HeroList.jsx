import { useMemo } from 'react';
import { getHeroesByPublisher } from '../helpers';
import { HeroItem } from './HeroItem';


export const HeroList = ({ publisher }) => {

    // console.log( publisher );
    const heroes = useMemo( () => getHeroesByPublisher( publisher ), [ publisher ] );

    return (
        
        <>
            <div className="row rows-cols-1 row-cols-md-3 g-3">
                {
                    heroes.map( hero => (

                        <HeroItem
                            key={ hero.id }
                            { ...hero }
                        />
                        // <li
                        //     key={ hero.id }
                        // >
                        //     { hero.superhero }
                        // </li>

                    ))
                }
            </div>
        </>

    )

}
