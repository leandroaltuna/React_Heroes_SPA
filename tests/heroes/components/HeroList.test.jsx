import { render, screen } from '@testing-library/react';
import { getHeroesByPublisher } from '../../../src/heroes/helpers';
import { HeroList } from '../../../src/heroes/components/HeroList';
import { MemoryRouter } from 'react-router-dom';


jest.mock( '../../../src/heroes/helpers' );


describe( 'Pruebas en <HeroList />', () => { 
    
    const publisher = 'DC Comics';

    test( 'Debe de mostrar info de los Heroes', () => { 
        
        const DC = [
            {
                id: 'dc-batman',
                superhero: 'Batman',
                publisher: 'DC Comics',
                alter_ego: 'Bruce Wayne',
                first_appearance: 'Detective Comics #27',
                characters: 'Bruce Wayne'
            },
            {
                id: 'dc-superman',
                superhero: 'Superman',
                publisher: 'DC Comics',
                alter_ego: 'Kal-El',
                first_appearance: 'Action Comics #1',
                characters: 'Kal-El'
            },
            {
                id: 'dc-flash',
                superhero: 'Flash',
                publisher: 'DC Comics',
                alter_ego: 'Jay Garrick',
                first_appearance: 'Flash Comics #1',
                characters: 'Jay Garrick, Barry Allen, Wally West, Bart Allen'
            },
        ];


        getHeroesByPublisher.mockReturnValue( DC );

        render(
            <MemoryRouter>
                <HeroList publisher={ publisher } />
            </MemoryRouter>
        );

        // screen.debug();
        const heroImg = screen.getAllByRole( 'img' );

        expect( heroImg.length ).toBe( 3 );
        
    });

    test( 'Si no hay heroes no debe mostrar nada', () => { 
        
        getHeroesByPublisher.mockReturnValue( [] );

        const { container } = render(
            <MemoryRouter>
                <HeroList publisher={ publisher } />
            </MemoryRouter>
        );

        // screen.debug();
        expect( container ).toMatchSnapshot();
        
    });
    
});