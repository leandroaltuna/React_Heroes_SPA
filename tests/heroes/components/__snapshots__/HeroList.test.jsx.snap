// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <HeroList /> Si no hay heroes no debe mostrar nada 1`] = `
<div>
  <div
    class="row rows-cols-1 row-cols-md-3 g-3"
  />
</div>
`;
